### ISSUE TITLE
Android 7.0

### ISSUE THREAD

**Issue by milesje** [AdAway/AdAway#794]
Fails to copy host file on Android 7.0 on Nexus6P.

**Comment by Snackzz** [comment]
Confirmed, I am having the same problem. I have just been downloading the hosts file directly and copy that into systems/etc.

**Comment by rfjord** [comment]
Same problem on Nexus 5X with 7.0, logcat of what I hope is the relevant bits:
```
08-26 12:43:22.038 12536 12536 E linker  : normalize_path - invalid input: "null", the input path should be absolute
08-26 12:43:22.038 12536 12536 W linker  : Warning: unable to normalize "null"
08-26 12:43:22.102 12547 12547 E linker  : normalize_path - invalid input: "null", the input path should be absolute
08-26 12:43:22.102 12547 12547 W linker  : Warning: unable to normalize "null"
08-26 12:43:22.173 12528 12528 W IntentService[A: type=1400 audit(0.0:164): avc: denied { read } for name="RootToolsMounts" dev="dm-0" ino=752195 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:system_data_file:s0 tclass=file permissive=0
08-26 12:43:22.178 10544 12528 E RootCommands: Exception
08-26 12:43:22.178 10544 12528 E RootCommands: java.io.FileNotFoundException: /data/local/RootToolsMounts (Permission denied)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at java.io.FileInputStream.open(Native Method)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at java.io.FileInputStream.<init>(FileInputStream.java:146)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at java.io.FileInputStream.<init>(FileInputStream.java:99)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at java.io.FileReader.<init>(FileReader.java:58)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at org.sufficientlysecure.rootcommands.Remounter.getMounts(Remounter.java:173)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at org.sufficientlysecure.rootcommands.Remounter.remount(Remounter.java:66)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at org.sufficientlysecure.rootcommands.Toolbox.remount(Toolbox.java:733)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at org.adaway.util.ApplyUtils.copyHostsFile(ApplyUtils.java:164)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at org.adaway.service.ApplyService.apply(ApplyService.java:427)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at org.adaway.service.ApplyService.doWakefulWork(ApplyService.java:102)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at com.commonsware.cwac.wakeful.WakefulIntentService.onHandleIntent(WakefulIntentService.java:100)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at android.os.Handler.dispatchMessage(Handler.java:102)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at android.os.Looper.loop(Looper.java:154)
08-26 12:43:22.178 10544 12528 E RootCommands: 	at android.os.HandlerThread.run(HandlerThread.java:61)
08-26 12:43:22.178 10544 12528 E AdAway  : Remounting as RW failed! Probably not a problem!
08-26 12:43:22.216 10544 12528 E AdAway  : Exception!
08-26 12:43:22.216 10544 12528 E AdAway  : org.adaway.util.CommandException
08-26 12:43:22.216 10544 12528 E AdAway  : 	at org.adaway.util.ApplyUtils.copyHostsFile(ApplyUtils.java:176)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at org.adaway.service.ApplyService.apply(ApplyService.java:427)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at org.adaway.service.ApplyService.doWakefulWork(ApplyService.java:102)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at com.commonsware.cwac.wakeful.WakefulIntentService.onHandleIntent(WakefulIntentService.java:100)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at android.os.Handler.dispatchMessage(Handler.java:102)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at android.os.Looper.loop(Looper.java:154)
08-26 12:43:22.216 10544 12528 E AdAway  : 	at android.os.HandlerThread.run(HandlerThread.java:61)
08-26 12:43:22.219 12562 12562 E linker  : normalize_path - invalid input: "null", the input path should be absolute
08-26 12:43:22.219 12562 12562 W linker  : Warning: unable to normalize "null"
08-26 12:43:22.293 12528 12528 W IntentService[A: type=1400 audit(0.0:165): avc: denied { read } for name="RootToolsMounts" dev="dm-0" ino=752195 scontext=u:r:untrusted_app:s0:c512,c768 tcontext=u:object_r:system_data_file:s0 tclass=file permissive=0
08-26 12:43:22.297 10544 12528 E RootCommands: Exception
08-26 12:43:22.297 10544 12528 E RootCommands: java.io.FileNotFoundException: /data/local/RootToolsMounts (Permission denied)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at java.io.FileInputStream.open(Native Method)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at java.io.FileInputStream.<init>(FileInputStream.java:146)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at java.io.FileInputStream.<init>(FileInputStream.java:99)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at java.io.FileReader.<init>(FileReader.java:58)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at org.sufficientlysecure.rootcommands.Remounter.getMounts(Remounter.java:173)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at org.sufficientlysecure.rootcommands.Remounter.remount(Remounter.java:66)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at org.sufficientlysecure.rootcommands.Toolbox.remount(Toolbox.java:733)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at org.adaway.util.ApplyUtils.copyHostsFile(ApplyUtils.java:191)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at org.adaway.service.ApplyService.apply(ApplyService.java:427)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at org.adaway.service.ApplyService.doWakefulWork(ApplyService.java:102)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at com.commonsware.cwac.wakeful.WakefulIntentService.onHandleIntent(WakefulIntentService.java:100)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at android.os.Handler.dispatchMessage(Handler.java:102)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at android.os.Looper.loop(Looper.java:154)
08-26 12:43:22.297 10544 12528 E RootCommands: 	at android.os.HandlerThread.run(HandlerThread.java:61)
08-26 12:43:22.297 10544 12528 E AdAway  : Remounting failed in finally! Probably not a problem!
08-26 12:43:22.298 10544 12528 E AdAway  : Exception: 
08-26 12:43:22.298 10544 12528 E AdAway  : org.adaway.util.CommandException
08-26 12:43:22.298 10544 12528 E AdAway  : 	at org.adaway.util.ApplyUtils.copyHostsFile(ApplyUtils.java:186)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at org.adaway.service.ApplyService.apply(ApplyService.java:427)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at org.adaway.service.ApplyService.doWakefulWork(ApplyService.java:102)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at com.commonsware.cwac.wakeful.WakefulIntentService.onHandleIntent(WakefulIntentService.java:100)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at android.app.IntentService$ServiceHandler.handleMessage(IntentService.java:67)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at android.os.Handler.dispatchMessage(Handler.java:102)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at android.os.Looper.loop(Looper.java:154)
08-26 12:43:22.298 10544 12528 E AdAway  : 	at android.os.HandlerThread.run(HandlerThread.java:61)
```

**Comment by rfjord** [comment]
There's a solution in an earlier issue for the previews: https://github.com/AdAway/AdAway/issues/770#issuecomment-242117387

**Comment by DaveCalaway** [comment]
@rfjord
Doesn't works... After I had applied the modifications, the app works without problem. But when I open the browser, all ads are still there. 
Nexus 5x 7.0

**Comment by rfjord** [comment]
@DaveCalaway Did you reboot?

**Comment by DaveCalaway** [comment]
Yep, I did. But now I retried the modifications and the console gave me this error.
![screenshot_20160828-135125](https://cloud.githubusercontent.com/assets/10884164/18033484/b27f3932-6d26-11e6-94db-30af2d925475.png)

**Comment by rfjord** [comment]
You are using a period (-orw.remount) where there should be a comma (orw,remount).

**Comment by DaveCalaway** [comment]
Oh sorry,  my mistake! But the Ads remain. ..

**Comment by Snackzz** [comment]
it should be said that if you are using Termux you need to make sure you are in the root directory type `cd /` after you `su`

**Comment by DaveCalaway** [comment]
@Snackzz done,  but the problem remain.

**Comment by haegar** [comment]
Even with the manually installed symlink in /system/etc/hosts to /data/data/hosts (and adaway writes into that file, and it exists with the correct content):

08-28 22:18:42.284  8461  8461 W netd    : type=1400 audit(0.0:328): avc: denied { open } for path="/data/data/hosts" dev="dm-0" ino=1489578 scontext=u:r:netd:s0 tcontext=u:object_r:system_data_file:s0 tclass=file permissive=0

(Most likely some selinux foo, the normal unix permissions look ok)

Nexus 5X with Android 7.0

**Comment by DaveCalaway** [comment]
This way works. 
https://github.com/AdAway/AdAway/issues/770#issuecomment-242997754

**Comment by 0-kaladin** [comment]
android 7 has made /system not writeable to a lot of devices where it was previously.  I have adaway working fine if you are systemless root and flash the systemless zip file for adaway and leave it pointed in stock location.

**Comment by KreAch3R** [comment]
You can try this apk here: https://github.com/AdAway/AdAway/issues/770#issuecomment-248972477

**Comment by sanjay900** [comment]
I created a pull request that should get the app fixed.
https://github.com/AdAway/AdAway/pull/807

**Comment by bobobo1618** [comment]
I've got this running on my Pixel using the recent root and I've verified that the hosts file is correctly applied (`cat /etc/hosts`), however I'm still seeing ads all over the system. Even simply running `ping <host>` in a shell shows that the original IP is still being resolved. Any ideas?

**Comment by 0-kaladin** [comment]
Yea. I think we need to update for the pixel. I didn't test yet but it
looks like from chainfires post that /system is now actually /system_root
Will have to investigate how to have AdAway detect that and maybe bind
mount hosts from there.

**Comment by bobobo1618** [comment]
If it helps, I tried `mount -o bind /data/data/hosts /system_root/etc/hosts` but no luck.

**Comment by 0-kaladin** [comment]
did you change adaway's target to /system_root/etc/hosts as well after that?

**Comment by bobobo1618** [comment]
No but after doing so, it works.

However AdAway complains that a symlink from my chosen location to `/system/etc/hosts` doesn't exist.

**Comment by bobobo1618** [comment]
According to [the XDA article for Pixel root](http://www.xda-developers.com/root-is-now-available-for-the-google-pixel-and-pixel-xl/), this should also work:

```
mkdir /su/etc
cp /system/etc/hosts /su/etc/hosts
echo "#!/su/bin/sush\nmount -o bind /su/etc/hosts /system/etc/hosts" > /su/su.d/50adaway
chmod 0700 /su/su.d/50adaway
```

Not sure how this compares as a solution but I can verify it also works.

**Comment by sanjay900** [comment]
@0-kaladin you want me to put together a test fix for pixel devices? Or do you want to do it

**Comment by 0-kaladin** [comment]
Please go right ahead! Thanks!

On Sun, Oct 30, 2016, 05:11 Sanjay Govind <notifications@github.com> wrote:

@0-kaladin <https://github.com/0-kaladin> you want me to put together a
test fix for pixel devices? Or do you want to do it

—
You are receiving this because you were mentioned.


Reply to this email directly, view it on GitHub
<https://github.com/AdAway/AdAway/issues/794#issuecomment-257140170>, or mute
the thread

**Comment by sanjay900** [comment]
@bobobo1618 Does this one work if you use /system/etc/hosts? 
[AdAway-debug.apk.gz](https://github.com/AdAway/AdAway/files/561134/AdAway-debug.apk.gz)

**Comment by p1r473** [comment]
Hi, I keep getting the error that AdAway can't copy my hosts file. But, my phone is rooted.

I tried to mount the system as RW in RootExplorer, and also in terminal, following method 2 here:
http://android.stackexchange.com/questions/110927/how-to-mount-system-rewritable-or-read-only-rw-ro

However, when I try to mount it, I get the error 
**mount: '/dev/block/bootdevice/by-name/system_b'->'/system': Device or resource busy**

The only thing I can think of is when I installed twrp, it asked whether I wanted to keep the system drive as read-only, and I clicked yes. Is that the problem? (Ref: http://www.androidpolice.com/2015/06/23/twrp-v2-8-7-0-introduces-a-read-only-install-option-for-creating-ota-friendly-system-images-and-more/)
I don't think this is the issue, as the referenced material seems to state it only keeps the system as read only while its being backed up or something...

My root works, and other rooted apps work too.

According to
https://www.xda-developers.com/root-is-now-available-for-the-google-pixel-and-pixel-xl/
It says I should be able to get AdAway working by doing
mount -o rw,remount /system_root

When I tried this command, I got no error in terminal, but AdAway failed with the same error.
